core/store/migrations/migration1597695690/migrate.go

```diff
@@ -0,0 +1,160 @@
+package migration1597695690
+
+import (
+	"github.com/jinzhu/gorm"
+)
+
+// Migrate creates the offchain_reporting_job_specs table
+func Migrate(tx *gorm.DB) error {
+	return tx.Exec(`
+        CREATE TABLE pipeline_specs (
+            id SERIAL PRIMARY KEY,
+            dot_dag_source TEXT NOT NULL,
+            created_at timestamptz NOT NULL
+        );
+
+        CREATE INDEX idx_pipeline_specs_created_at ON pipeline_specs USING BRIN (created_at);
+
+        CREATE TABLE pipeline_task_specs (
+            id SERIAL PRIMARY KEY,
+            dot_id TEXT,
+            pipeline_spec_id INT NOT NULL REFERENCES pipeline_specs (id) ON DELETE CASCADE,
+            type TEXT NOT NULL,
+            json jsonb NOT NULL,
+            index INT NOT NULL DEFAULT 0,
+            successor_id INT REFERENCES pipeline_task_specs (id),
+            created_at timestamptz NOT NULL
+        );
+
+        CREATE INDEX idx_pipeline_task_specs_pipeline_spec_id ON pipeline_task_specs (pipeline_spec_id);
+        CREATE INDEX idx_pipeline_task_specs_successor_id ON pipeline_task_specs (successor_id);
+        CREATE INDEX idx_pipeline_task_specs_created_at ON pipeline_task_specs USING BRIN (created_at);
+
+        CREATE TABLE pipeline_runs (
+            id BIGSERIAL PRIMARY KEY,
+            pipeline_spec_id BIGINT NOT NULL REFERENCES pipeline_specs (id) ON DELETE CASCADE,
+            created_at timestamptz NOT NULL
+            -- NOTE: Could denormalize here with finished_at/output/error of last task_run if that proves necessary for performance
+        );
+
+        CREATE INDEX idx_pipeline_runs_pipeline_spec_id ON pipeline_runs (pipeline_spec_id);
+        CREATE INDEX idx_pipeline_runs_created_at ON pipeline_runs USING BRIN (created_at);
+
+        CREATE TABLE pipeline_task_runs (
+            id BIGSERIAL PRIMARY KEY,
+            dot_id TEXT,
+            pipeline_run_id BIGINT NOT NULL REFERENCES pipeline_runs (id) ON DELETE CASCADE,
+            output JSONB,
+            error TEXT,
+            pipeline_task_spec_id BIGINT NOT NULL REFERENCES pipeline_task_specs (id) ON DELETE CASCADE,
+            index INT NOT NULL DEFAULT 0,
+            created_at timestamptz NOT NULL,
+            finished_at timestamptz,
+            CONSTRAINT chk_pipeline_task_run_fsm CHECK (
+                error IS NULL AND output IS NULL AND finished_at IS NULL
+                OR
+                error IS NULL AND output IS NOT NULL AND finished_at IS NOT NULL
+                OR
+                output IS NULL AND error IS NOT NULL AND finished_at IS NOT NULL
+            )
+        );
+
+        CREATE OR REPLACE FUNCTION notifyJobRunCompletion() RETURNS TRIGGER AS $_$
```

Comment:
Careful! Triggers can be unexpectedly expensive. We'll need to do profiling on this

Reply:
Sure happy to, but this was your idea originally no?